server/src/middleware/mod.rs

```diff
@@ -33,6 +33,211 @@ mod host_filter;
 /// Proxy `GET /path` to internal RPC methods.
 mod proxy_get_request;
 
+use std::{net::SocketAddr, sync::Arc};
+
 pub use authority::*;
 pub use host_filter::*;
 pub use proxy_get_request::*;
+
+use http::{HeaderMap, Uri};
+use tower::layer::util::{Identity, Stack};
+use tower::layer::LayerFn;
+use tower::util::Either;
+
+use jsonrpsee_core::server::{
+	BoundedSubscriptions, MethodCallback, MethodResponse, MethodSink, Methods, SubscriptionState,
+};
+use jsonrpsee_core::traits::IdProvider;
+use jsonrpsee_types::error::{reject_too_many_subscriptions, ErrorCode};
+use jsonrpsee_types::{ErrorObject, Request};
+
+/// The transport protocol used to send or receive a call or request.
+#[derive(Debug, Copy, Clone)]
+pub enum TransportProtocol {
+	/// HTTP transport.
+	Http,
+	/// WebSocket transport.
+	WebSocket,
+}
+
+impl std::fmt::Display for TransportProtocol {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		let s = match self {
+			Self::Http => "http",
+			Self::WebSocket => "websocket",
+		};
+
+		write!(f, "{s}")
+	}
+}
+
+/// Context to a JSON-RPC call.
+#[derive(Debug, Clone)]
+pub struct Context {
+	/// Transport protocol.
+	pub transport: TransportProtocol,
+	/// Remote addr.
+	pub remote_addr: SocketAddr,
+	/// Connection id.
+	pub conn_id: usize,
+	/// HTTP Headers.
+	pub headers: HeaderMap,
+	/// URI.
+	pub uri: Uri,
+}
+
+/// Configuration of the RpcService.
+#[derive(Clone, Debug)]
+pub(crate) enum RpcServiceCfg {
+	/// The server supports only calls.
+	OnlyCalls,
+	/// The server supports both method calls and subscriptions.
+	CallsAndSubscriptions {
+		bounded_subscriptions: BoundedSubscriptions,
+		sink: MethodSink,
+		id_provider: Arc<dyn IdProvider>,
+		_pending_calls: tokio::sync::mpsc::Sender<()>,
+	},
+}
+
+/// JSON-RPC service middleware.
+#[derive(Clone, Debug)]
+pub struct RpcService {
+	conn_id: usize,
+	methods: Methods,
+	max_response_body_size: usize,
+	cfg: RpcServiceCfg,
+}
+
+impl RpcService {
+	/// Create a new service with doesn't support subscriptions.
+	pub(crate) fn new(methods: Methods, max_response_body_size: usize, conn_id: usize, cfg: RpcServiceCfg) -> Self {
+		Self { methods, max_response_body_size, conn_id, cfg }
+	}
+}
+
+/// Similar to `tower::Service` but specific for jsonrpsee and
+/// doesn't requires `&mut self` for performance reasons.
+#[async_trait::async_trait]
+pub trait RpcServiceT<'a> {
```

Comment:
There is also an alternative to do which doesn't requrire boxing the future as the async trait does:

```rust
pub trait Service {
    /// The future response value.
    type Future<'cx>: Future<Output = MethodResponse>>
    where
        Self: 'cx,
        Request: 'cx;

    /// Process the request and return the response asynchronously.
    fn call(&self, req: Request<'_>) -> Self::Future<'_>;
}
```

Reply:
This would allow more a little better performance especially if there are a bunch of middlewares, and our logging one could also then be as light weight as possible, so IMO it's probably worth doing this even if it makes the trait more annoying (but people can just set the type to `Box<dyn Future<Output = Foo> + 'static>` if they want to keep it simple anyway I think)

Reply:
Our stuff requires still `Send + 'static` and not sure if it's worth all the complexity with such Service trait